File: backend/document.py
import user
import json
import flask

from google.appengine.ext import ndb  # pylint: disable=E0401,E0611

import validation
from storage import Document, User

APP = flask.Blueprint('doc', __name__)


def get_documents_list(query):
    doc_titles = query.fetch()
    docs = []
    for doc in doc_titles:
        docs.append(json.loads(doc.content))
    return docs


@APP.route('/init', methods=['POST', 'GET'])
def init_docs():
    usr = user.get_user()
    if not usr or not usr.isAdmin:
        flask.abort(403)
    query = Document.query(Document.public == True)
    keys = query.fetch(keys_only=True)
    ndb.delete_multi(keys)
    data = json.load(open('catalog.json'))
    docs = []
    for item in data:
        doc = Document()
        doc.owner = ndb.Key(User, usr.key.id())
        doc.title = item[u'title']
        doc.type = item[u'type']
        doc.public = True
        doc.content = json.dumps(item)
        docs.append(doc)
    ndb.put_multi(docs)
    return '', 200


@APP.route('/list', methods=['POST', 'GET'])
def list_docs():
    query = Document.query(Document.public == True)
    docs = get_documents_list(query)
    uid = user.get_uid()
    if uid:
        query = Document.query(
            ndb.AND(Document.public == False,
                    Document.owner == ndb.Key(User, uid)))
        docs += get_documents_list(query)
    return flask.jsonify(docs)


@APP.route('/save', methods=['POST'])
def save():
    usr = user.get_user()
    if not usr:
        flask.abort(403)
    data = flask.request.get_json()
    typ = data.get(u'type', None)
    if typ:
        doc = Document.query(
            ndb.AND(Document.type == typ,
                    Document.owner == usr.key)).get()
    else:
        doc = Document()
        doc.owner = ndb.Key(User, usr.key.id())

    doc.title = data[u'title']
    doc.type = data[u'type']
    doc.public = data[u'public']
    doc.content = data[u'content']
    doc.put()

    return flask.jsonify({
        'docid': doc.key.id()
    })


@APP.route('/delete', methods=['POST'])
def delete():
    usr = user.get_user()
    if not usr:
        flask.abort(403)
    data = flask.request.get_json()
    typ = data.get(u'type', None)
    if not typ:
        flask.abort(400)
    doc = Document.query(
        ndb.AND(Document.type == typ,
                Document.owner == usr.key)).get()
    if not doc:
        flask.abort(400)
    if doc.owner.id() != usr.key.id():
        flask.abort(401)
    doc.key.delete()
    return 'Success.', 200

File: backend/src/document.js
'use strict';

const express = require('express');
const bodyParser = require('body-parser');
const docData = require('./storage/document');
const userData = require('./storage/user');
const user = require('./user');

const router = express.Router();

// Automatically parse request body as JSON
router.use(bodyParser.json());

// Use the oauth middleware to automatically get the user's profile
// information and expose login/logout URLs to templates.
router.use(user.template);

/**
 * POST /api/document/list
 *
 * List documents.
 */
router.post('/list', (req, res) => {
  let data = req.body;
  let category = data.category;

  let promises = [docData.read({public: true, category})];

  if (req.user && req.user.uid) {
    let owner = userData.key(req.user.uid);
    promises.push(docData.read({public: false, category, owner}));
  }

  Promise.all(promises).then(results => {
    let docs = [];
    results.forEach(result => docs = docs.concat(result));
    let contents = docs.map(doc => JSON.parse(doc.content));
    return contents;
  })
  .then(contents => {
    return res.json(contents);
  })
  .catch(e => {
    console.error(e.name + ': ' + e.message);
    return res.status(400).end();
  });
});

/**
 * POST /api/document/get
 *
 * List documents.
 */
router.post('/get/:identifier/:owner?', (req, res) => {
  let identifier = req.params.identifier;
  let owner = undefined;
  if (req.params.owner) {
    owner = userData.key(req.params.owner);
  }
  docData.read({identifier, owner}).then(docs => {
    if (docs.length == 0) {
      throw 'Requested doc not found.';
    };
    return res.send(docs[0].content);
  })
  .catch(e => {
    console.error(e.name + ': ' + e.message);
    return res.status(400).end();
  });
});

/**
 * POST /api/document/save
 *
 * Create a new document.
 */
router.post('/save', (req, res) => {
  if (!req.user) {
    return res.status(403).end();
  }

  let data = req.body;
  let {title, identifier, category, public: public_} = data;

  if (!req.user.isAdmin) {
    public_ = false;
  }

  if (!identifier) {
    return res.status(400).end();
  }

  let content = JSON.stringify(data);
  let owner = userData.key(req.user.uid);

  docData.update({
    title, identifier, category, public: public_, owner, content
  });

  return res.status(200).end();
});


/**
 * POST /api/document/save
 *
 * Create a new document.
 */
router.post('/delete', (req, res) => {
  if (!req.user) {
    return res.status(403).end();
  }

  let data = req.body;
  let identifier = data.identifier;

  if (!identifier) {
    return res.status(400).end();
  }

  let owner = userData.key(req.user.uid);

  docData.findAndRemove({identifier, owner}).then(() => {
    return res.status(200).end();
  });
});

/**
 * Errors on "/api/document/*" routes.
 */
router.use((err, req, res, next) => {
  // Format error and forward to generic error handler for logging and
  // responding to the request
  err.response = {
    message: err.message,
    internalCode: err.code
  };
  next(err);
});

module.exports = {
  router
};
